featbit/featbit
# ✨ feat: allow retrieving projects using service token

A Developer must have an environment id in order to use any of the OpenApi functions that are availabe with ManageFeatureFlag or ManageSegment permissions. Environments are part of projects, but only personal access tokens and logged in user accounts are able retrieve information from the Project Controller.  This PR allows a Service Token to get a project or a list of all projects in an organization.  This is need to assisted developers who wish to access flag or segment using a service token.

An IOperationFilter called OrganizationHeaderParameter was also added to allow developers to test this functionality via Swagger.

This is an update to the API and the UI. 

To test the ui:
1. navigate to the UI
2. login with credentials that have the ability to create access tokens
3. expand "Integrations" on the left navigation bar
4. click "Access Tokens"
5. Click Add
6. Name the token
7. Select "Service" in the Type dropdown
8. Observe that "Project " appears in the Permissions section
9. Expand Project 
10. Observe that both All, and "Manage project (ManageProject)" actions are selected
11. Observe that the resource is "project/*"
12. Click save
13. Record the token
14. Click Ok.


  
To test the api:
1. Run the API in development mode
2. navigate to http://localhost:5000/swagger
3. Click Authorize
4. Use your newly created Service Token to authenticate
5. Enter "Project" in the filter box or scroll to it
6. Expand GET /api/v1/projects
7. Click Try it Out
8. Enter your Organization Id, if you don't knwo it look for it in mongodb.
9. Click Execute 
10. Observe that a list of project is displayed
11. Expand GET /api/v1/projects/{projectId}
12. Click Try It out
13. Enter a project id from the list displayed in step 15
14. Click execute
15. Observe that project information is returned
16. Note and enviroment id from the environment section of the result
16. Replace "Project" with "FeatureFlag"
18. expand GET /api/va/envs/{envId}/feature-flags
19. Click Try it out
20. Enter the envId noted in step 21
21. Click Execute
22. Observe that Flags and Flag details for the environment are returned

Alternate way to test the API using curl
```
curl --location 'http://localhost:5000/api/v1/projects' \
--header 'Authorization: <Access Token> ' \
--header 'Content-Type: application/json' \
--header 'Organization: <OrganizationId>'
```
